tlsutil/pki.go

```diff
@@ -0,0 +1,61 @@
+package tlsutil
+
+import (
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"strings"
+)
+
+type TLSBundle struct {
+	Certificate tls.Certificate
+	PublicKey   crypto.PublicKey
+	PrivateKey  crypto.PrivateKey
+	CertPool    *x509.CertPool
+}
+
+func FromPKI(certificate string, privateKey string) (*TLSBundle, error) {
+	trimC := strings.Trim(certificate, " ")
+	trimK := strings.Trim(privateKey, " ")
+	cert, err := tls.X509KeyPair([]byte(trimC), []byte(trimK))
+	if err != nil {
+		// Try possible flattening in PKI credential
+		normalizedCert := stringToPEMFormat(trimC)
+		normalizedKey := stringToPEMFormat(trimK)
+		cert, err = tls.X509KeyPair([]byte(normalizedCert), []byte(normalizedKey))
+		if err != nil {
+			return nil, fmt.Errorf("cannot load certificate, error %w", err)
+		}
+		return FromPKICert(&cert)
+	}
+	return FromPKICert(&cert)
+}
+
+func FromPKICert(cert *tls.Certificate) (*TLSBundle, error) {
+	var err error
+	// Create Leaf form of certificate
+	cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return nil, fmt.Errorf("cannot load ceritifcate, error: %w", err)
+	}
+	config := &TLSBundle{
+		Certificate: *cert,
+		PublicKey:   cert.Leaf.PublicKey,
+		CertPool:    x509.NewCertPool(),
+		PrivateKey:  cert.PrivateKey,
+	}
+	config.CertPool.AddCert(cert.Leaf)
+	return config, nil
+}
+
+// Re-format PEM string which might lost format during the data conversions
+func stringToPEMFormat(pem string) string {
```

Comment:
I'm not entirely sure I understand why this is necessary, is this some issue caused by using `openssl` for the generation ?

Reply:
Previously, I had all kinds of broken PEM formats coming from different sources like JSON/YAML :) It's just some safety tooling for me :)